libr/io/ioutils.c

```diff
@@ -0,0 +1,68 @@
+/* radare - LGPL - Copyright 2017 - condret */
+
+#include <r_io.h>
+#include <r_util.h>
+#include <r_types.h>
+
+// when io.va is true this checks if the highest priorized map at this
+// offset has the same or high permissions set. When there is no map it
+// check for the current desc permissions and size.
+// when io.va is false it only checks for the desc
+R_API bool r_io_is_valid_real_offset(RIO* io, ut64 offset, int hasperm) {
+	RIOMap* map;
+	if (!io) {
+		return false;
+	}
+	if (io->va && (map = r_io_map_get (io, offset))) {
+		return ((map->flags & hasperm) == hasperm);
+	}
+	if (!io->desc) {
+		return false;
+	}
+	if (r_io_desc_size (io->desc) <= offset) {
+		return false;
+	}
+	return ((io->desc->flags & hasperm) == hasperm);
```

Comment:
personal preference

My personal preference is to make `*_real_offset` simpler and don't bother degrading to `[0, r_io_desc_)` when no mapping is available.